flow/display_list.cc

```diff
@@ -0,0 +1,1248 @@
+#undef DRAW_ATLAS_HAS_COLORS_ARRAY
+#undef DRAW_ATLAS_NO_CULLING_ARGS
+#undef DRAW_ATLAS_NO_CULLING_INIT
+#undef DRAW_ATLAS_NO_CULLING_FIELDS
+#undef DRAW_ATLAS_NO_CULLING_P_ARG
+#undef DRAW_ATLAS_HAS_CULLING_ARGS
+#undef DRAW_ATLAS_HAS_CULLING_INIT
+#undef DRAW_ATLAS_HAS_CULLING_FIELDS
+#undef DRAW_ATLAS_HAS_CULLING_P_ARG
+
+// 4 byte header + ptr aligned payload uses 12 bytes rounde up to 16
+// (4 bytes unused)
+struct DrawSkPictureOp final : DLOp {
+  static const auto kType = DisplayListOpType::DrawSkPicture;
+
+  DrawSkPictureOp(sk_sp<SkPicture> picture) : picture(std::move(picture)) {}
+
+  sk_sp<SkPicture> picture;
+
+  void dispatch(Dispatcher& dispatcher) const {
+    dispatcher.drawPicture(picture);
+  }
+};
+
+// 4 byte header + ptr aligned payload uses 12 bytes rounde up to 16
+// (4 bytes unused)
+struct DrawDisplayListOp final : DLOp {
+  static const auto kType = DisplayListOpType::DrawDisplayList;
+
+  DrawDisplayListOp(const sk_sp<DisplayList> display_list)
+      : display_list(std::move(display_list)) {}
+
+  sk_sp<DisplayList> display_list;
+
+  void dispatch(Dispatcher& dispatcher) const {
+    dispatcher.drawDisplayList(display_list);
+  }
+};
+
+struct DrawTextBlobOp final : DLOp {
+  static const auto kType = DisplayListOpType::DrawTextBlob;
+
+  DrawTextBlobOp(const sk_sp<SkTextBlob> blob, SkScalar x, SkScalar y)
+      : x(x), y(y), blob(std::move(blob)) {}
+
+  const SkScalar x;
+  const SkScalar y;
+  const sk_sp<SkTextBlob> blob;
+
+  void dispatch(Dispatcher& dispatcher) const {
+    dispatcher.drawTextBlob(blob, x, y);
+  }
+};
+
+// struct DrawShadowRecOp final : DLOp {
+//   static const auto kType = DisplayListOpType::DrawShadowRec;
+//
+//   DrawShadowRecOp(const SkPath& path, const SkDrawShadowRec& rec)
+//       : path(path), rec(rec) {}
+//
+//   const SkPath path;
+//   const SkDrawShadowRec rec;
+//
+//   void dispatch(Dispatcher& dispatcher) const {
+//     dispatcher.drawShadowRec(path, rec);
+//   }
+// };
+
+// 4 byte header + 28 byte payload packs evenly into 32 bytes
+struct DrawShadowOp final : DLOp {
+  static const auto kType = DisplayListOpType::DrawShadow;
+
+  DrawShadowOp(const SkPath& path,
+               SkColor color,
+               SkScalar elevation,
+               bool occludes)
+      : color(color), elevation(elevation), occludes(occludes), path(path) {}
+
+  const SkColor color;
+  const SkScalar elevation;
+  const bool occludes;
+  const SkPath path;
+
+  void dispatch(Dispatcher& dispatcher) const {
+    dispatcher.drawShadow(path, color, elevation, occludes);
+  }
+};
+
+void DisplayList::computeBounds() {
+  DisplayListBoundsCalculator calculator(boundsCull_);
+  dispatch(calculator);
+  bounds_ = calculator.getBounds();
+}
+
+void DisplayList::dispatch(Dispatcher& dispatcher,
+                           uint8_t* ptr,
+                           uint8_t* end) const {
+  while (ptr < end) {
+    auto op = (const DLOp*)ptr;
+    ptr += op->size;
```

Comment:
General question here:

Why are we using raw basically untyped pointers and pointer arithmetic? 

Are we really saving enough by avoiding virtual dispatch to make it worth having this surface for memory corruption/lack of type checking?

Reply:
See the SkLiteDL reference in the description. It is Skia's recommendation for how to implement a DisplayList and it is in use by both Android and Chrome in just this form.

Reply:
(let's discuss this here: https://github.com/flutter/engine/pull/26928#discussion_r658254626)